tests/test_socket_server.py

```diff
@@ -0,0 +1,180 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import asyncio
+import json
+import logging
+import typing
+import unittest
+
+from lsst.ts import tcpip
+from lsst.ts.ess import common
+
+logging.basicConfig(
+    format="%(asctime)s:%(levelname)s:%(name)s:%(message)s", level=logging.DEBUG
+)
+
+TIMEOUT = 5
+"""Standard timeout in seconds."""
+
+
+class SocketServerTestCase(unittest.IsolatedAsyncioTestCase):
+    async def asyncSetUp(self) -> None:
+        self.ctrl = None
+        self.writer = None
+        self.mock_ctrl = None
+        self.name = "MockDevice"
+        self.srv = common.SocketServer(
+            name="EssCommonTest", host="0.0.0.0", port=0, simulation_mode=1
+        )
+        mock_command_handler = common.MockCommandHandler(
+            callback=self.srv.write, simulation_mode=1, name=self.name
+        )
+        self.srv.set_command_handler(mock_command_handler)
+
+        self.log = logging.getLogger(type(self).__name__)
+
+        await self.srv.start_task
+        self.assertTrue(self.srv.server.is_serving())
+        self.reader, self.writer = await asyncio.open_connection(
+            host=tcpip.LOCAL_HOST, port=self.srv.port
+        )
+
+    async def asyncTearDown(self) -> None:
+        await self.srv.disconnect()
+        if self.writer:
+            self.writer.close()
+            await self.writer.wait_closed()
+        await self.srv.exit()
+
+    async def read(self) -> typing.Dict[str, typing.Any]:
+        """Read a string from the reader and unmarshal it.
+
+        Returns
+        -------
+        data : `dict`
+            A dictionary with objects representing the string read.
+        """
+        read_bytes = await asyncio.wait_for(
+            self.reader.readuntil(tcpip.TERMINATOR), timeout=TIMEOUT
+        )
+        data = json.loads(read_bytes.decode())
+        return data
+
+    async def write(self, **data: typing.Dict[str, typing.Any]) -> None:
+        """Write the data appended with a tcpip.TERMINATOR string.
+
+        Parameters
+        ----------
+        data:
+            The data to write.
+        """
+        st = json.dumps({**data})
+        assert self.writer is not None
+        self.writer.write(st.encode() + tcpip.TERMINATOR)
+        await self.writer.drain()
+
+    async def test_disconnect(self) -> None:
+        self.assertTrue(self.srv.connected)
+        await self.write(command=common.Command.DISCONNECT, parameters={})
+        # Give time to the socket server to clean up internal state and exit.
+        await asyncio.sleep(0.5)
```

Comment:
If your server has a task you can await then this will be more robust and more efficient if you await that task here (with a time limit).

The underlying OneClientServer has a `done_task` that might be appropriate.

Reply:
I gave this a try but I am probably doing something wrong because I cannot get it to work. I'll need some help with this.

Reply:
We talked about this and agreed to make it a separate ticket. If your SocketServer close does anything asynchronous after telling OneClientServer to disconnect then using OneClientServer's done_task will not be sufficient because it will be marked done before SocketServer is fully closed.